custom_components/elasticsearch/__init__.py

```diff
@@ -145,13 +143,14 @@ async def async_migrate_entry(
         config_entry.version = 3
 
     if config_entry.version == 3:
-        newOptions = {**config_entry.options}
+        new = get_merged_config(config_entry)
 
         # Check the configured options for the index_mode
-        if CONF_INDEX_MODE not in newOptions:
-            newOptions[CONF_INDEX_MODE] = INDEX_MODE_LEGACY
+        if CONF_INDEX_MODE not in new:
+            new[CONF_INDEX_MODE] = INDEX_MODE_LEGACY
 
-        hass.config_entries.async_update_entry(config_entry, options=newOptions, version=4)
+        config_entry.data = {**new}
+        config_entry.version = 4

```

Comment:
The existing tests were failing to handle the call to `async_update_entry`. It appeared that the config flow was in an invalid state. Rather than debug that, I opted to adjust your migration to match the format of the existing migrations, since that's proven to work.

I'll try again once we update the required HA version to once which supports the `version` argument.